ironfish/src/rpc/adapters/tcpAdapter.ts

```diff
@@ -0,0 +1,207 @@
+import { IpcError, IpcRequest, IpcResponse, IpcStream } from './ipcAdapter'
+
+// Message type that node-ipc client sends over a TCP socket
+type IncomingNodeIpc = {
+  type: 'message'
+  data: IpcRequest
+}
+
+// Message type that node-ipc client listens for over a TCP socket
+type OutgoingNodeIpc =
+  | { type: 'message'; data: IpcResponse }
+  | { type: 'malformedRequest'; data: IpcError }
+  | { type: 'stream'; data: IpcStream }
+
+export const IncomingNodeIpcSchema: yup.ObjectSchema<IncomingNodeIpc> = yup
+  .object({
+    type: yup.string().oneOf(['message']).required(),
+    data: yup
+      .object({
+        mid: yup.number().required(),
+        type: yup.string().required(),
+        data: yup.mixed().notRequired(),
+      })
+      .required(),
+  })
+  .required()
+
+export class TcpAdapter implements IAdapter {
+  logger: Logger
+  host: string
+  port: number
+  server: net.Server | null = null
+  router: Router | null = null
+  namespaces: ApiNamespace[]
+
+  constructor(
+    host: string,
+    port: number,
+    logger: Logger = createRootLogger(),
+    namespaces: ApiNamespace[],
+  ) {
+    this.host = host
+    this.port = port
+    this.logger = logger.withTag('tcpadapter')
+    this.namespaces = namespaces
+  }
+
+  start(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.server = net.createServer((socket) => this.onClientConnection(socket))
+
+      this.server.on('error', (err) => {
+        reject(err)
+      })
+
+      this.server.listen(
+        {
+          host: this.host,
+          port: this.port,
+          exclusive: true,
+        },
+        () => {
+          resolve()
+        },
+      )
+    })
+  }
+
+  stop(): Promise<void> {
+    this.logger.debug(`tcpAdapter stopped: ${this.host}:${this.port}`)
+    return Promise.resolve()
+  }
+
+  attach(server: RpcServer): void {
+    this.router = server.getRouter(this.namespaces)
+  }
+
+  unattach(): void {
+    this.router = null
+  }
+
+  onClientConnection(socket: net.Socket): void {
+    socket.on('data', (data) => {
+      this.onClientData(socket, data).catch((err) => this.logger.error(err))
+    })
+    socket.on('close', () => {
+      //TODO: possibly keep track of ongoing reqeusts on this connection and
+      // cancel all of them. This seems like an optimization but not a requirement
+      this.logger.debug(`client connection closed: ${this.host}:${this.port}`)
+    })
+    socket.on('error', (error: Error) => {
+      this.logger.debug(`${this.host}:${this.port} has error : ${error.message}`)
+    })
+  }
+
+  async onClientData(socket: net.Socket, data: Buffer): Promise<void> {
+    const dataString = data.toString('utf8').trim()
+    const result = await YupUtils.tryValidate(IncomingNodeIpcSchema, dataString)
+
+    if (result.error) {
+      this.logger.log(
+        `${this.host}:${this.port} Error parsing message : ${result.error.message} raw: ${dataString} `,
+      )
```

Comment:
I know were getting heavily into preference, but if there's one style thing I could advocate for our entire code base that I've adopted from bcoin that I think is very good is not nesting things, and bailing on error cases when there is no more execution flow. So return in this if statement, and unindent the rest of the code.